README.md
# [logger-request](https://github.com/hex7c0/logger-request)

[![NPM version](https://img.shields.io/npm/v/logger-request.svg)](https://www.npmjs.com/package/logger-request)
[![Linux Status](https://img.shields.io/travis/hex7c0/logger-request.svg?label=linux)](https://travis-ci.org/hex7c0/logger-request)
[![Windows Status](https://img.shields.io/appveyor/ci/hex7c0/logger-request.svg?label=windows)](https://ci.appveyor.com/project/hex7c0/logger-request)
[![Dependency Status](https://img.shields.io/david/hex7c0/logger-request.svg)](https://david-dm.org/hex7c0/logger-request)
[![Coveralls](https://img.shields.io/coveralls/hex7c0/logger-request.svg)](https://coveralls.io/r/hex7c0/logger-request)

HTTP request logger middleware for [nodejs](http://nodejs.org/), standalone logger and even more!

Save logs to file, show to console or both, to MongoDb, etc...

Look at [`logger-request-cli`](https://github.com/hex7c0/logger-request-cli) for Parser

## Installation

Install through NPM

```bash
npm install logger-request
```
or
```bash
git clone git://github.com/hex7c0/logger-request.git
```

## API

inside expressjs project
```js
var logger = require('logger-request');

var app = require('express')();

app.use(logger({
  filename: 'foo.log',
}));
```

### logger(options)

#### options

 - `transports` - **Array** Array of [winston transports](https://github.com/winstonjs/winston/blob/master/docs/transports.md) *(default "false")* 
 - `filename` - **String** If string, filename of the logfile to write output to *(default "false")*
 - `console` - **Boolean** If true, it displays log to console *(default "false")*
 - `standalone` - **Boolean** If true, return logger function instead of callback *(default "false")*
 - `deprecated` - **Boolean** Flag for write log after `res.end()`(true) instead of default `listener`(false) *(default "false")*
 - `functions` - **String** Use module like a function without `next` callback *(default "false")*
 - `winston` - **Object** Setting for selected transports
  - `logger` - **String** Logger option related to [`winston`](https://github.com/flatiron/winston#working-with-multiple-loggers-in-winston) *(default "logger-request")*
  - `level` - **String** Level of messages that this transport should log *(default "info")*
  - `silent` - **Boolean** Flag indicating whether to suppress output *(default "false")*
  - `colorize` - **Boolean** Flag indicating if we should colorize output *(default "false")*
  - `timestamp` - **Boolean|Function** Flag indicating if we should prepend output with timestamps *(default "true")*. If function is specified, its return value will be used instead of timestamps
  - `maxsize` - **Number** Max size in bytes of the logfile, if the size is exceeded then a new file is created *(default "8388608" [8Mb])*
  - `maxFiles` - **Number** Limit the number of files created when the size of the logfile is exceeded *(default "no limit")*
  - `json` - **Boolean** If true, messages will be logged as JSON *(default "true")*
  - `raw` - **Boolean** If true, raw messages will be logged to console *(default "false")*
  - `...` - **Mixed** Extra settings
 - `custom` - **Object** Setting for customization of logs
  - `pid` - **Boolean** Flag for `process.pid` *(default "disabled")*
  - `bytesReq` - **Boolean** Flag for `req.socket.bytesRead` *(default "disabled")*
  - `bytesRes` - **Boolean** Flag for `req.socket._bytesDispatched` *(default "disabled")*
  - `referer` - **Boolean** Flag for `req.headers['referer']` *(default "disabled")*
  - `auth` - **Boolean** Flag for [`basic-authentication`](https://github.com/hex7c0/basic-authentication) *(default "disabled")*
  - `transfer` - **Boolean** Flag for [`transfer-rate`](https://github.com/hex7c0/transfer-rate) *(default "disabled")*
  - `agent` - **Boolean** Flag for `req.headers['user-agent']` *(default "disabled")*
  - `lang` - **Boolean** Flag for `req.headers['accept-language']` *(default "disabled")*
  - `cookie` - **Boolean** Flag for `req.cookies` *(default "disabled")*
  - `headers` - **Boolean** Flag for `req.headers` *(default "disabled")*
  - `version` - **Boolean** Flag for `req.httpVersionMajor` *(default "disabled")*
  - `callback` - **Function** Flag for using callback function *(default "disabled")*

## Examples

Take a look at my [examples](examples)

### [License GPLv3](LICENSE)


History.md
v3.7.0 / 2016-09-03
==================

  * Rotate log each day
  * Add `daily` (options)
  * Tested against `node`@6

v3.6.1 / 2016-03-06
==================

  * Tested against `node`@5.3
  * Update `winston`@2.2.0

v3.6.0 / 2015-12-04
==================

  * Tested against `node`@5
  * Update `winston`@2.1.1
  * Update `basic-authentication`@1.6.2
  * Update `transfer-rate`@1.1.3

v3.5.0 / 2015-10-26
==================

  * Tested against `node`@4
  * Update `winston`@1.1.1
  * Update `basic-authentication`@1.6.1
  * Update `transfer-rate`@1.1.2

v3.4.0 / 2015-08-27
==================

  * Tested against `iojs`@3
  * Small improvements
  * Update `basic-authentication`@1.6.0
  * Update `transfer-rate`@1.1.1

v3.3.5 / 2015-06-30
==================

  * Update `winston`@1.0.1

v3.3.4 / 2015-06-07
==================

  * Fix SPDX license
  * Update `on-finished`@2.2.1

v3.3.3 / 2015-05-05
==================

  * Update `on-finished`@2.2.1

v3.3.2 / 2015-04-11
==================

  * Update `transfer-rate`@1.1.0
  * Update `winston`@1.0.0

v3.3.1 / 2015-02-28
==================

  * Ignore `/coverage` for npm
  * Update `basic-authentication`@1.5.12

v3.3.0 / 2015-02-15
==================

  * Add `transports` (options)
  * `filename` (options) default value is `false`
  * `coveralls` test

v3.2.8 / 2015-02-05
==================

  * `windows` test
  * `iojs` test
  * Update `basic-authentication`@1.5.11
  * Update `transfer-rate`@1.0.6
  * Update `winston`@0.9.0

v3.2.7 / 2015-01-02
==================

  * Update `transfer-rate`@1.0.5

v3.2.6 / 2014-12-30
==================

  * Remove "x-forwarded-for" ip

v3.2.5 / 2014-12-23
==================

  * Use `hrtime` before/after all
  * Update `on-finished`@2.2.0

v3.2.4 / 2014-12-19
==================

  * Decrease decimal unit to .2 in `response`

v3.2.3 / 2014-12-02
==================

  * Callback function inside "custom" object

v3.2.2 / 2014-11-21
==================

  * Update `basic-authentication`@1.5.9

v3.2.1 / 2014-11-15
==================

  * Small improvement

v3.2.0 / 2014-11-10
==================

  * Add `transfer` (options)
  * Using `transfer-rate`@1.0.4

v3.1.2 / 2014-11-06
==================

  * Fix for Node 0.11
  * Update `winston`@0.8.3

  v3.1.1 / 2014-11-03
==================

  * Strict If

v3.1.0 / 2014-11-01
==================

  * Don't check `next` callback
  * Add `function` (options)

v3.0.13 / 2014-10-26
==================

  * Update `basic-authentication`@1.5.8

v3.0.12 / 2014-10-26
==================

  * `jshint`

v3.0.11 / 2014-10-23
==================

  * Update `on-finished`@2.1.1

v3.0.10 / 2014-10-08
==================

  * Update `winston`@0.8.1

v3.0.9 / 2014-10-04
==================

  * Remove jsdoc
  * `package.json` min

v3.0.8 / 2014-09-29
==================

  * Little fix

v3.0.7 / 2014-09-18
==================

  * Update `winston`@0.8.0

v3.0.5 / 2014-09-01
==================

  * New `grunt-endline`@0.1.0

v3.0.4 / 2014-09-01
==================

  * Update `basic-authentication`@1.5.5

v3.0.3 / 2014-08-23
==================

  * Performance tips
  * Update `basic-authentication`@1.5.4

v3.0.2 / 2014-08-17
==================

  * Change `req` to `res` on-finished callback
  * Update `on-finished`@2.1.0

v3.0.1 / 2014-08-16
==================

  * Save IP for multiple chunks
  * Using `on-finished`@2.0.0

v3.0.0 / 2014-08-14
==================

  * "deprecated" (options)
  * Using `finished`@1.2.2
  * Use event listener instead of rewrite `res.end`
   * You can use old method with "deprecated" flag

v2.2.5 / 2014-08-11
==================

  * Update README.md
  * Update `basic-authentication`@1.5.2

v2.2.3 / 2014-08-04
==================

  * Doc update

v2.2.2 / 2014-07-31
==================

  * Fix "bytesRes" and "bytesReq"
  * Add "headers" (options)

v2.2.0 / 2014-07-27
==================

  * Test "node": ">=0.10.0" only
  * Using task runner `grunt`
  * Using test framework `mocha`
  * Testing script will be put inside "test/"
  * ".npmignore" more aggressive
  * `uglify` compiles
  * `jsdoc` documentation

v2.1.1 / 2014-07-25
==================

  * Update `basic-authentication`@1.4.0

v2.1.0 / 2014-07-25
==================

  * Rewrite for multiple require

v2.0.3 / 2014-07-24
==================

  * Fix `options.custom` if void

v2.0.2 / 2014-07-23
==================

  * `res.end` correct return Boolean
  * Fix `options.winston.silent` for display only console out and not file

v2.0.0 / 2014-07-22
==================

  * Add `custom` object. Now can change log output
  * Change _API_. This setting now, are under `winston` object
   * `logger`
   * `level`
   * `silent`
   * `colorize`
   * `timestamp`
   * `maxsize`
   * `maxFiles`
   * `json`
   * `raw`

v1.3.0 / 2014-07-19
==================

  * _log_ Rename "byte" to "bytesRes", bytes read
  * _log_ New "bytesReq", bytes dispatched
  * _log_ New "__filename", who wrote log
  * Now working even as a function

v1.2.0 / 2014-07-17
==================

  * Don"t store "cookie" anymore
  * Minor var to function
  * Change MINOR version, due critical issue with node 0.11

v1.1.9 / 2014-07-16
==================

  * Improve performance (remove closure)
  * Write log, after sending all stuff to client

v1.1.8 / 2014-07-09
==================

  * Fix critical issue with node 0.10.29 (callback)

v1.1.7 / 2014-06-29
==================

  * Remove anonymous function
  * Update [`express`](https://github.com/visionmedia/express) @ 4.4.5

v1.1.5 / 2014-06-18
==================

  * Clean code

v1.1.4 / 2014-06-15
==================

  * Small improvements
  * Print output bytes

v1.1.3 / 2014-06-08
==================

  * Improve closures functions

v1.1.2 / 2014-06-08
==================

  * Various fixes

v1.1.1 / 2014-06-05
==================

  * Best callback function

v1.1.0 / 2014-05-27
==================

  * Fix status code
  * JsDoc improvement

v1.0.10 / 2014-05-25
==================

  * Fix "logger" arg


index.js
'use strict';
/**
 * @file logger-request main
 * @module logger-request
 * @subpackage main
 * @version 3.4.0
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * functions
 */
/**
 * output builder
 * 
 * @private
 * @function info
 * @param {Object} my - user options
 * @return {Function}
 */
function __info(my) {

  var promise = Array.call();
  if (my.pid) {
    promise.push([ 'pid', function() {

      return process.pid;
    } ]);
  }
  if (my.bytesReq) {
    promise.push([ 'bytesReq', function(req) {

      return req.socket.bytesRead;
    } ]);
  }
  if (my.bytesRes) {
    promise.push([ 'bytesRes', function(req) {

      return req.socket._bytesDispatched;
    } ]);
  }
  if (my.referer) {
    promise.push([ 'referer', function(req) {

      return req.headers.referer || req.headers.referrer;
    } ]);
  }
  if (my.auth) {
    var mod0 = require('basic-authentication')({
      legacy: true
    });
    promise.push([ 'auth', function(req) {

      return mod0(req).user;
    } ]);
  }
  if (my.transfer) {
    var mod1 = require('transfer-rate')();
    promise.push([ 'transfer', function(req) {

      return mod1(req, req.start);
    } ]);
  }
  if (my.agent) {
    promise.push([ 'agent', function(req) {

      return req.headers['user-agent'];
    } ]);
  }
  if (my.lang) {
    promise.push([ 'lang', function(req) {

      return req.headers['accept-language'];
    } ]);
  }
  if (my.cookie) {
    promise.push([ 'cookie', function(req) {

      return req.cookies;
    } ]);
  }
  if (my.headers) {
    promise.push([ 'headers', function(req) {

      return req.headers;
    } ]);
  }
  if (my.version) {
    promise.push([ 'version', function(req) {

      return req.httpVersionMajor + '.' + req.httpVersionMinor;
    } ]);
  }
  if (my.callback) {
    promise.push([ 'callback', function(req) {

      return my.callback(req);
    } ]);
  }
  var l = promise.length;

  if (l === 0) {
    /**
     * standard logger output
     * 
     * @function io
     * @param {Object} req - client request
     * @param {Integer} statusCode - response status code
     * @return {Object}
     */
    return function io(req, statusCode, end) {

      var diff = end[0] * 1e3 + end[1] * 1e-6;
      return {
        ip: req.remoteAddr || req.ip,
        method: req.method,
        status: statusCode,
        url: req.url,
        response: diff.toFixed(2)
      };
    };
  }
  /**
   * logger output after info builder
   * 
   * @function io
   * @param {Object} req - client request
   * @param {Integer} statusCode - response status code
   * @param {Array} end - high resolution time
   * @return {Object}
   */
  return function io(req, statusCode, end) {

    var diff = end[0] * 1e3 + end[1] * 1e-6;
    var out = {
      ip: req.remoteAddr || req.ip,
      method: req.method,
      status: statusCode,
      url: req.url,
      response: diff.toFixed(2)
    };
    for (var i = 0; i < l; ++i) {
      var p = promise[i];
      out[p[0]] = p[1](req);
    }
    return out;
  };
}

/**
 * function wrapper for multiple require
 * 
 * @function wrapper
 * @param {Function} log - logging function
 * @param {Object} my - parsed options
 * @param {Function} io - extra function
 * @return {Function}
 */
function wrapper(log, my, io) {

  var finished = require('on-finished');

  /**
   * end of job (closures). Get response time and status code
   * 
   * @function finale
   * @param {Object} req - client request
   * @param {Integer} statusCode - response status code
   * @param {Array} start - high resolution time
   * @return {Null}
   */
  function finale(req, statusCode, start) {

    req.start = start;
    return log(my.logger, io(req, statusCode, process.hrtime(start)));
  }

  if (my.deprecated) {
    /**
     * logging all route
     * 
     * @deprecated
     * @function deprecated
     * @param {Object} req - client request
     * @param {Object} res - response to client
     * @param {Function} next - next callback
     */
    return require('util').deprecate(function deprecated(req, res, next) {

      var start = process.hrtime();
      req.remoteAddr = req.ip;
      if (res._headerSent) { // function
        finale(req, res.statusCode, start); // after res.end()
      } else { // middleware
        var buffer = res.end;
        /**
         * middle of job. Set right end function (closure)
         * 
         * @function
         * @param {String} chunk - data sent
         * @param {String} encoding - data encoding
         * @return {Boolean}
         */
        res.end = function end(chunk, encoding) {

          res.end = buffer;
          var b = res.end(chunk, encoding);
          // res.end(chunk,encoding,finale) // callback available only
          // with node 0.11
          finale(req, res.statusCode, start); // write after sending
          // all stuff, instead of callback
          return b;
        };
      }

      return next ? next() : null;
    }, '`logger-request` option is deprecated');
  }

  if (my.functions) {
    /**
     * logging all route without next callback
     * 
     * @function logging
     * @param {Object} req - client request
     * @param {Object} res - response to client
     * @return {Null}
     */
    return function logging(req, res) {

      req.remoteAddr = req.ip;
      return finale(req, res.statusCode, process.hrtime());
    };
  }

  /**
   * logging all route
   * 
   * @function logging
   * @param {Object} req - client request
   * @param {Object} res - response to client
   * @param {Function} next - continue routes
   * @return {Null}
   */
  return function logging(req, res, next) {

    var start; // closure
    req.remoteAddr = req.ip;

    finished(res, function() {

      return finale(req, res.statusCode, start);
    });

    start = process.hrtime();
    return next();
  };
}

/**
 * option setting
 * 
 * @exports logger
 * @function logger
 * @param {Object} opt - various options. Check README.md
 * @return {Function|Object}
 */
function logger(opt) {

  var winston = require('winston');

  var options = opt || Object.create(null);
  var my = {
    filename: options.filename || false,
    daily: Boolean(options.daily),
    transports: Array.isArray(options.transports) ? options.transports : []
  };
  if (Boolean(options.deprecated)) {
    my.deprecated = true;
  } else if (Boolean(options.functions)) {
    my.functions = true;
  }

  // winston
  var optional = options.winston || Object.create(null);
  // default option for File transport and Console
  optional.logger = String(optional.logger || 'logger-request');
  optional.level = String(optional.level || 'info');
  optional.timestamp = optional.timestamp || true;
  optional.maxsize = Number(optional.maxsize) || 8388608;
  optional.maxFiles = Number(optional.maxFiles) || null;
  optional.json = optional.json === false ? false : true;
  optional.raw = optional.raw === false ? false : true;
  my.logger = optional.logger;

  var log = new winston.Logger(); // without transport

  if (my.filename) {
    optional.filename = require('path').resolve(String(my.filename));
    if (my.daily) {
      optional.prepend = true;
      log.add(require('winston-daily-rotate-file'), optional);
    } else {
      log.add(winston.transports.File, optional);
    }
  }

  if (Boolean(options.console)) {
    log.add(winston.transports.Console, optional);
  }

  for (var i = 0, ii = my.transports.length; i < ii; ++i) {
    log.add(my.transports[i], optional);
  }
  log = log[optional.level]; // extract logger level function

  if (Boolean(options.standalone)) {
    return log;
  }

  // custom
  options = options.custom || Object.create(null);

  return wrapper(log, my, __info({
    pid: Boolean(options.pid),
    bytesReq: Boolean(options.bytesReq),
    bytesRes: Boolean(options.bytesRes),
    referer: Boolean(options.referer),
    auth: Boolean(options.auth),
    transfer: Boolean(options.transfer),
    agent: Boolean(options.agent),
    lang: Boolean(options.lang),
    cookie: Boolean(options.cookie),
    headers: Boolean(options.headers),
    version: Boolean(options.version),
    callback: typeof options.callback == 'function' ? options.callback : false
  }));
}
module.exports = logger;
